src/swresample_stubs.c

```diff
@@ -0,0 +1,628 @@
+  value out_vector;
+  int out_size;
+  int out_buf_size;
+
+  int (*set_in_data)(swr_t *, value);
+  void (*alloc_out_vector)(swr_t *, int);
+  int (*convert)(swr_t *, uint8_t **, int, int);
+};
+
+#define Swr_val(v) (*(swr_t**)Data_custom_val(v))
+
+
+static int set_in_data_frame(swr_t *swr, value in_vector)
+{
+  AVFrame *frame = Frame_val(in_vector);
+  int nb_channels = av_frame_get_channels(frame);
+    
+  if (nb_channels != swr->in_nb_channels) {
+    snprintf(error_msg, ERROR_MSG_SIZE, "Swresample failed to convert %d channels : %d channels were expected", nb_channels, swr->in_nb_channels);
+    Raise(EXN_FAILURE, error_msg);
+  }
+
+  if (frame->format != swr->in_sample_fmt) {
+    snprintf(error_msg, ERROR_MSG_SIZE, "Swresample failed to convert %s sample format : %s sample format were expected", av_get_sample_fmt_name(frame->format), av_get_sample_fmt_name(swr->in_sample_fmt));
+    Raise(EXN_FAILURE, error_msg);
+  }
+
+  swr->in_data = frame->extended_data;
+  return frame->nb_samples;
+}
+
+static int set_in_data_string(swr_t *swr, value in_vector)
+{
+  swr->in_data[0] = (uint8_t*)String_val(in_vector);
+  return caml_string_length(in_vector) / swr->in_bytes_per_samples;
+}
+
+static int set_in_data_planar_string(swr_t *swr, value in_vector)
+{
+  value str = Field(in_vector, 0);
+  int str_len = caml_string_length(str);
+
+  swr->in_data[0] = (uint8_t*)String_val(str);
+  int in_nb_samples = str_len / swr->in_bytes_per_samples;
+    
+  for (int i = 1; i < swr->in_nb_channels; i++) {
+    str = Field(in_vector, i);
+    
+    if(str_len != caml_string_length(str)) {
+      snprintf(error_msg, ERROR_MSG_SIZE, "Swresample failed to convert channel %d's %lu bytes : %d bytes were expected", i, caml_string_length(str), str_len);
+      Raise(EXN_FAILURE, error_msg);
+    }
+    swr->in_data[i] = (uint8_t*)String_val(str);
+  }
+  return in_nb_samples;
+}
+
+static int set_in_data_float_array(swr_t *swr, value in_vector)
+{
+  swr->in_data[0] = (uint8_t*)in_vector;
+  return Wosize_val(in_vector) / (Double_wosize * swr->in_nb_channels);
+}
+
+static int set_in_data_planar_float_array(swr_t *swr, value in_vector)
+{
+  value fa = Field(in_vector, 0);
+  int fa_len = Wosize_val(fa);
+
+  swr->in_data[0] = (uint8_t*)fa;
+  int in_nb_samples = fa_len / Double_wosize;
+    
+  for (int i = 1; i < swr->in_nb_channels; i++) {
+    fa = Field(in_vector, i);
+    
+    if(fa_len != Wosize_val(fa)) {
+      snprintf(error_msg, ERROR_MSG_SIZE, "Swresample failed to convert channel %d's %lu bytes : %d bytes were expected", i, Wosize_val(fa), fa_len);
+      Raise(EXN_FAILURE, error_msg);
+    }
+    swr->in_data[i] = (uint8_t*)fa;
+  }
+  return in_nb_samples;
+}
+
+static int set_in_data_ba(swr_t *swr, value in_vector)
+{
+  swr->in_data[0] = Caml_ba_data_val(in_vector);
+  return Caml_ba_array_val(in_vector)->dim[0] / swr->in_nb_channels;
+}
+
+static int set_in_data_planar_ba(swr_t *swr, value in_vector)
+{
+  value ba = Field(in_vector, 0);
```

Comment:
I have a doubt. The macro CAMLparam being applied to the in_vector value in the ocaml_swresample_convert function, I thought it would be sufficient to protect all values referenced by it. If this is not the case, the use of the macro CAMLlocal protects only the last Bigarray of the channel array.
Moreover, the problem exists for all functions set_in_data_planar_* isn't it?
Do you know a detailed documentation about Ocaml/C bindings?